embassy-stm32/src/flash/common.rs

```diff
@@ -162,6 +162,35 @@ impl FlashRegion {
     }
 }
 
+impl embedded_storage::nor_flash::ErrorType for Flash<'_> {
+    type Error = Error;
+}
+
+impl embedded_storage::nor_flash::ReadNorFlash for Flash<'_> {
+    const READ_SIZE: usize = 1;
+
+    fn read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), Self::Error> {
+        self.blocking_read(offset, bytes)
+    }
+
+    fn capacity(&self) -> usize {
+        FLASH_SIZE
+    }
+}
+
+impl embedded_storage::nor_flash::NorFlash for Flash<'_> {
+    const WRITE_SIZE: usize = 8;
+    const ERASE_SIZE: usize = 2048;
```

Comment:
Ahh, you are absolutely right. And that might have been the reason why @rmja did not implement this? `WRITE_SIZE` is easy, but what should `ERASE_SIZE` be here, as that depends on the region?

Reply:
On some chips there's sectors with different sizes. For example some H7s have [16, 16, 16, 16, 64, 128, 128, ...] kbs. If you want to use the smaller sectors individually through `embedded-storage` (for e.g. embassy-boot stuff) you need different instances for the different erase sizes. Hence the idea of doing the "regions" thing in #1297.

We discussed on Matrix that it loses the ability of addressing the entire flash at once using a single `embedded-storage` instance. We concluded that it's wanted, but it was fine to not have it for now, it could be added in the future.

It is possible to do, it just requires "uniformizing" the sectors to the biggest size. For example in the H7 if you erase the first 128k sector through embedded-storage, it'd erase the first 5 sectors [16, 16, 16, 16, 64] which happen to add up to 128k.

Reply:
so to be clear, adding an `embedded-storage` impl for `Flash` like you did sounds good to me. All it needs is setting write_size/erase_size based on the chip metadata, and setting erase_size to the max if there's mutliple.

This way there's the "native" API which supports non-uniform sectors, then there's the whole-flash e-s impl, then there's the individual regions e-s impls.

Reply:
Yeah, i got that part.

My question is more along the lines of `setting erase_size to the max if there's mutliple`, where would i obtain these in within embassy?

Reply:
it comes from the metadata: https://github.com/embassy-rs/embassy/blob/master/embassy-stm32/build.rs#L175

you can probably generate some consts from build.rs then use them from the impl.

Reply:
Ahh, perfect :+1: 
Thanks